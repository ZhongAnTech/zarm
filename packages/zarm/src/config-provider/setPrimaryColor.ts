import Color from 'color';
import type { ConfigProviderProps } from './PropsType';

const setPrimaryColor = (color: ConfigProviderProps['primaryColor']) => {
  document.documentElement.style.setProperty('--theme-primary', color);
  document.documentElement.style.setProperty(
    '--theme-primary-dark',
    Color(color).darken(0.05).string(),
  );
  document.documentElement.style.setProperty(
    '--theme-primary-lighter',
    Color(color).lightness(95).string(),
  );
  document.documentElement.style.setProperty(
    '--button-primary-shadow-color',
    Color(color).alpha(0.3).string(),
  );
};

export default setPrimaryColor;
