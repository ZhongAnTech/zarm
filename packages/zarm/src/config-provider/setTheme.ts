import type { Theme } from './PropsType';

export const themes = {
  '--theme-primary-lighter': '#303030',
  '--color-text': 'rgba(255, 255, 255, 0.85)',
  '--color-text-inverse': 'rgba(255, 255, 255, 0.8)',
  '--color-text-placeholder': 'rgba(255, 255, 255, 0.3)',
  '--color-text-disabled': '#666',
  '--border-color': '#3a3b3d',
  '--opacity-mask': '0.7',
  '--calendar-background': 'transparent',
  '--calendar-week-background': '#303030',
  '--picker-background': '#000',
  '--picker-header-background': '#1b1c1e',
  '--picker-mask-background-start': 'rgba(0, 0, 0, 0.4)',
  '--picker-mask-background-end': 'rgba(0, 0, 0, 0.8)',
  '--stack-picker-background': '#000',
  '--stack-picker-shadow': 'none',
} as const;

const setTheme = (value: Theme) => {
  document.body.setAttribute('data-theme', value);
  Object.keys(themes).forEach((key) => {
    value === 'dark'
      ? document.documentElement.style.setProperty(key, themes[key])
      : document.documentElement.style.removeProperty(key);
  });
};

export default setTheme;
