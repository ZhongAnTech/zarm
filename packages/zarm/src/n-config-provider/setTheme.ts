import type { ConfigProviderProps } from './interface';

const themes = {
  '--background-color': '#393939',
  '--background-active': '#222',
  '--theme-primary-lighter': '#303030',
  '--color-text': 'rgba(255, 255, 255, 0.85)',
  '--color-text-inverse': 'rgba(255, 255, 255, 0.8)',
  '--color-text-placeholder': 'rgba(255, 255, 255, 0.3)',
  '--color-text-disabled': '#666',
  '--border-color': '#1a1a1a',
  '--opacity-mask': '0.7',
  '--stack-picker-background': '#000',
  '--stack-picker-shadow': 'none',

  '--za-background-color': '#1c1c1e',
  '--za-background-active': '#363738',
  '--za-theme-primary-lighter': '#303030',
  '--za-calendar-background': 'transparent',
  '--za-calendar-week-background': '#303030',
  '--za-calendar-month-scroll-background': 'rgba(0, 0, 0, .5)',
  '--za-color-text': 'rgba(255, 255, 255, 0.85)',
  '--za-color-text-inverse': 'rgba(255, 255, 255, 0.8)',
  '--za-color-text-placeholder': 'rgba(255, 255, 255, 0.3)',
  '--za-color-text-disabled': '#666',
  '--za-border-color': '#3a3b3d',
  '--za-opacity-mask': '0.7',
  '--za-activity-indicator-path-color': '#3a3b3d',
  '--za-button-background': '#393939',
  '--za-button-border-color': 'transparent',
  '--za-button-text-color': 'rgba(255, 255, 255, 0.85)',
  '--za-button-active-background': 'rgba(255, 255, 255, 0.2)',
  '--za-button-active-border-color': 'transparent',
  '--za-button-active-text-color': '#fff',
  '--za-badge-border-color': 'transparent',
  '--za-list-item-background': '#1c1c1e',
  '--za-list-item-active-background': '#363738',
  '--za-list-item-arrow-color': '#666',
  '--za-list-item-separator-color': 'rgba(84, 84, 88, 0.65)',
  '--za-list-item-info-text-color': 'rgba(235, 235, 245, 0.6)',
  '--za-list-item-after-text-color': 'rgba(235, 235, 245, 0.6)',
  '--za-loading-background': '#2b2c2d',
  '--za-collapse-arrow-color': '#666',
  '--za-collapse-arrow-disabled-color': '#333',
  '--za-checkbox-widget-background': 'transparent',
  '--za-checkbox-widget-border-color': 'rgb(72, 72, 74)',
  '--za-checkbox-disabled-widget-background': '#333',
  '--za-checkbox-disabled-widget-border-color': 'rgba(72, 72, 74, 0.7)',
  '--za-checkbox-disabled-text-color': 'rgba(255, 255, 255, 0.3)',
  '--za-checkbox-disabled-marker-color': 'rgba(255, 255, 255, 0.2)',
  '--za-radio-widget-background': 'transparent',
  '--za-radio-widget-border-color': 'rgb(72, 72, 74)',
  '--za-radio-disabled-widget-background': '#333',
  '--za-radio-disabled-widget-border-color': 'rgba(72, 72, 74, 0.7)',
  '--za-radio-disabled-text-color': 'rgba(255, 255, 255, 0.3)',
  '--za-radio-disabled-marker-color': 'rgba(255, 255, 255, 0.2)',
  '--za-input-clear-icon-color': '#8e8e93',
  '--za-keyboard-background': '#323234',
  '--za-keyboard-item-background': '#434343',
  '--za-keyboard-item-box-shadow': '0 1px 0 #000',
  '--za-switch-background': 'rgba(120, 120, 128, 0.32)',
  '--za-stepper-input-background': '#000',
  '--za-panel-header-color': 'rgba(235, 235, 245, 0.6)',
  '--za-panel-body-background': '#1c1c1e',
  '--za-progress-background': 'var(--za-border-color)',
  '--za-picker-background': '#000',
  '--za-picker-header-background': '#1b1c1e',
  '--za-picker-mask-background-start': 'rgba(0, 0, 0, 0.4)',
  '--za-picker-mask-background-end': 'rgba(0, 0, 0, 0.8)',
  '--za-picker-item-selected-background': 'rgba(116, 116, 128, 0.18)',
  '--za-modal-background': '#2b2c2d',
  '--za-modal-close-color': 'rgba(255, 255, 255, 0.3)',
  '--za-modal-close-active-color': 'rgba(255, 255, 255, 0.65)',
  '--za-nav-bar-background': '#1b1c1e',
  '--za-slider-line-background': '#3a3b3d',
  '--za-slider-dot-border-color': '#3a3b3d',
  '--za-search-bar-input-background': 'rgba(118, 118, 128, 0.24)',
  '--za-search-bar-input-placeholder-color': '#8e8e93',
  '--za-search-bar-icon-color': '#8e8e93',
  '--za-action-sheet-item-background': '#222222',
  '--za-action-sheet-cancel-background': '#2c2c2e',
  '--za-rate-color': '#393939',
  '--za-tabbar-background': '#1b1c1e',
  '--za-toast-background': '#2b2c2d',
  '--za-tooltip-background': '#5b5c60',
} as const;

const setTheme = (value: ConfigProviderProps['theme']) => {
  document.body.setAttribute('data-theme', value!);
  Object.keys(themes).forEach((key) => {
    value === 'dark'
      ? document.documentElement.style.setProperty(key, themes[key])
      : document.documentElement.style.removeProperty(key);
  });
};

export default setTheme;
