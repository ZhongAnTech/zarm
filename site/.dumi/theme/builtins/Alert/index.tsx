import type { AlertProps } from 'antd';
import { Alert } from 'antd';
import React from 'react';

const MdAlert: React.FC<AlertProps> = ({ style, ...props }) => (
  <Alert {...props} style={{ margin: '24px 0', ...style }} />
);

export default MdAlert;
