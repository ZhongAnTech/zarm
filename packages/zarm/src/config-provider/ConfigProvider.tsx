import React, { PureComponent, createContext } from 'react';
import type { Context } from 'react';
import type { ConfigProviderProps, Locale } from './PropsType';
import setTheme from './setTheme';
import setPrimaryColor from './setPrimaryColor';

const defaultConfig: ConfigProviderProps = {
  locale: ({} as unknown) as Locale,
  theme: 'light',
  primaryColor: '#00bc70',
};

export const LocaleContext: Context<Locale> = createContext(defaultConfig.locale);
export const ConfigContext: Context<ConfigProviderProps> = createContext(defaultConfig);

let runTimeLocale: Locale;
const changeRunTimeLocale = (locale: Locale) => {
  runTimeLocale = locale;
};

export const getRunTimeLocale = () => runTimeLocale;

export default class ConfigProvider extends PureComponent<ConfigProviderProps, {}> {
  static defaultProps = defaultConfig;

  componentDidMount() {
    this.update();
  }

  componentDidUpdate() {
    this.update();
  }

  update() {
    const { locale, theme, primaryColor } = this.props;
    changeRunTimeLocale(locale);
    setTheme(theme);
    setPrimaryColor(primaryColor);
  }

  render() {
    const { locale, theme, primaryColor, children } = this.props;
    return (
      <ConfigContext.Provider value={{ locale, theme, primaryColor }}>
        {React.Children.only(children)}
      </ConfigContext.Provider>
    );
  }
}
