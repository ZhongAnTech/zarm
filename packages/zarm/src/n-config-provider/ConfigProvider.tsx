import * as React from 'react';
import type { Context } from 'react';
import type { ConfigProviderProps } from './interface';
import defaultLocaleData from './locale/zh_CN';
import setCssVars, { defaultCssVars } from './setCssVars';
import setPrimaryColor from './setPrimaryColor';
import setTheme from './setTheme';

const defaultConfig: ConfigProviderProps = {
  prefixCls: 'za',
  locale: defaultLocaleData,
  theme: 'light',
  safeIphoneX: false,
  cssVars: defaultCssVars,
};

export const ConfigContext: Context<ConfigProviderProps> = React.createContext(defaultConfig);

let runtimeConfigContext: ConfigProviderProps;

const changeRuntimeConfigContext = (props: ConfigProviderProps) => {
  runtimeConfigContext = props;
};

const ConfigProvider: React.FC<ConfigProviderProps> = (props) => {
  const { children, cssVars, primaryColor, theme, ...rest } = props;

  changeRuntimeConfigContext(props);

  React.useEffect(() => {
    primaryColor && setPrimaryColor(primaryColor);
  }, [primaryColor]);

  React.useEffect(() => {
    theme && setTheme(theme);
  }, [theme]);

  const newChildren = setCssVars(children, cssVars);

  return (
    <ConfigContext.Provider value={rest}>{React.Children.only(newChildren)}</ConfigContext.Provider>
  );
};

export const RuntimeConfigProvider: React.FC = ({ children }) => {
  const props = React.useRef(runtimeConfigContext);

  React.useEffect(() => {
    props.current = runtimeConfigContext;
  }, [runtimeConfigContext]);

  return <ConfigProvider {...props?.current}>{children}</ConfigProvider>;
};

ConfigProvider.displayName = 'ConfigProvider';
ConfigProvider.defaultProps = defaultConfig;

export default ConfigProvider;
