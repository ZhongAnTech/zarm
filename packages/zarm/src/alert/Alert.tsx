import * as React from 'react';
import classnames from 'classnames';
import type { BaseAlertProps } from './interface';
import Modal from '../modal';
import { ConfigContext } from '../n-config-provider';
import ModalButton from '../modal/ModalButton';

export interface AlertProps extends BaseAlertProps {
  className?: string;
  style?: React.CSSProperties;
}

const Alert: React.FC<AlertProps> = (props) => {
  const { className, shape, content, cancelText, onCancel, ...rest } = props;
  const { prefixCls: globalPrefixCls, locale } = React.useContext(ConfigContext);
  const prefixCls = `${globalPrefixCls}-alert`;
  const classes = classnames(prefixCls, className, {
    [`${prefixCls}--${shape}`]: !!shape,
  });

  return (
    <div className={classes}>
      <Modal
        {...rest}
        footer={
          <ModalButton
            className={`${prefixCls}__button`}
            loadingClassName={`${prefixCls}__button--loading`}
            onClick={onCancel}
          >
            {cancelText || locale?.Alert.cancelText}
          </ModalButton>
        }
      >
        {content}
      </Modal>
    </div>
  );
};

Alert.displayName = 'Alert';

Alert.defaultProps = {
  animationType: 'zoom',
  shape: 'radius',
};

export default Alert;
