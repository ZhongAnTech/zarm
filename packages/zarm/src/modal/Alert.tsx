import * as React from 'react';
import Alert from '../alert';
import renderToContainer from '../utils/renderToContainer';
import { RuntimeConfigProvider } from '../n-config-provider/ConfigProvider';
import type { AlertProps } from '../alert';
import type { ContainerType } from '../utils/dom';

const alert = (props: Omit<AlertProps, 'visible'>) => {
  return new Promise((resolve) => {
    let unmount = () => {};
    const Wrapper = () => {
      const [visible, setVisible] = React.useState(false);
      React.useEffect(() => {
        setVisible(true);
      }, []);

      return (
        <RuntimeConfigProvider>
          <Alert
            {...props}
            visible={visible}
            mountContainer={false}
            onCancel={async () => {
              const close = await props.onCancel?.();
              if (close === false) return;
              setVisible(false);
              resolve(null);
            }}
            afterClose={() => {
              props.afterClose?.();
              unmount();
            }}
          />
        </RuntimeConfigProvider>
      );
    };

    unmount = renderToContainer(props.mountContainer as ContainerType, <Wrapper />);
  });
};

export default alert;
