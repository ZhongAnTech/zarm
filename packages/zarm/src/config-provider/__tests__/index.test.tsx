/* eslint-disable @typescript-eslint/no-var-requires */
import React from 'react';
import { mount, render, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Alert from '../../alert';
import localeCN from '../locale/zh_CN';
import localeEN from '../locale/en_US';
import setTheme, { themes } from '../setTheme';
import setPrimaryColor from '../setPrimaryColor';
import { ConfigContext } from '../ConfigProvider';

class TestComponent extends React.Component {
  static contextType = ConfigContext;

  render() {
    const { theme } = this.context;
    return <div>{theme}</div>;
  }
}

describe('setTheme', () => {
  afterEach(() => {
    jest.restoreAllMocks();
    document.body.removeAttribute('data-theme');
    document.documentElement.style.cssText = '';
  });
  it('should set dark theme', () => {
    expect.assertions(4 + Object.keys(themes).length);
    const setAttributeSpy = jest.spyOn(document.body, 'setAttribute');
    const setPropertySpy = jest.spyOn(document.documentElement.style, 'setProperty');
    setTheme('dark');
    expect(setAttributeSpy).toBeCalledWith('data-theme', 'dark');
    expect(document.body.getAttribute('data-theme')).toEqual('dark');
    Object.keys(themes).forEach((key) => {
      expect(setPropertySpy).toBeCalledWith(key, themes[key]);
    });
    expect(document.documentElement.style).toHaveLength(Object.keys(themes).length);
    expect(document.documentElement.style.cssText).toMatchSnapshot();
  });

  it('should set light theme', () => {
    expect.assertions(4 + Object.keys(themes).length);
    const setAttributeSpy = jest.spyOn(document.body, 'setAttribute');
    const removePropertySpy = jest.spyOn(document.documentElement.style, 'removeProperty');
    setTheme('light');
    expect(setAttributeSpy).toBeCalledWith('data-theme', 'light');
    expect(document.body.getAttribute('data-theme')).toBe('light');
    Object.keys(themes).forEach((key) => {
      expect(removePropertySpy).toBeCalledWith(key);
    });
    expect(document.documentElement.style).toHaveLength(0);
    expect(document.documentElement.style.cssText).toMatchSnapshot();
  });
});

describe('setPrimaryColor', () => {
  afterEach(() => {
    jest.restoreAllMocks();
    document.documentElement.style.cssText = '';
  });
  it('should set primary theme color', () => {
    const setPropertySpy = jest.spyOn(document.documentElement.style, 'setProperty');
    setPrimaryColor('#00bc70');
    expect(setPropertySpy).toBeCalledWith('--theme-primary', '#00bc70');
    expect(setPropertySpy).toBeCalledWith(
      '--theme-primary-dark',
      'hsl(155.70000000000005, 100%, 35%)',
    );
    expect(setPropertySpy).toBeCalledWith(
      '--theme-primary-lighter',
      'hsl(155.70000000000005, 100%, 95%)',
    );
    expect(setPropertySpy).toBeCalledWith(
      '--button-primary-shadow-color',
      'rgba(0, 188, 112, 0.3)',
    );
    expect(document.documentElement.style.cssText).toMatchInlineSnapshot(
      `"--theme-primary: #00bc70; --theme-primary-dark: hsl(155.70000000000005, 100%, 35%); --theme-primary-lighter: hsl(155.70000000000005, 100%, 95%); --button-primary-shadow-color: rgba(0, 188, 112, 0.3);"`,
    );
  });
});

describe('ConfigProvider', () => {
  describe('snapshot', () => {
    let ConfigProvider: typeof import('../index').default;
    beforeEach(() => {
      ConfigProvider = require('../index').default;
    });
    it('renders correctly with zh_CN', () => {
      const wrapper = render(
        <ConfigProvider locale={localeCN}>
          <Alert />
        </ConfigProvider>,
      );
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('renders correctly with en_US', () => {
      const wrapper = render(
        <ConfigProvider locale={localeEN}>
          <Alert />
        </ConfigProvider>,
      );
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('behaviour', () => {
    let mod: typeof import('../ConfigProvider');
    let ConfigProvider: typeof mod.default;
    const mSetTheme = jest.fn();
    const mSetPrimaryColor = jest.fn();

    beforeEach(() => {
      jest.resetModules();
      jest.doMock('../setTheme', () => {
        return {
          __esModule: true,
          default: mSetTheme,
        };
      });
      jest.doMock('../setPrimaryColor', () => {
        return {
          __esModule: true,
          default: mSetPrimaryColor,
        };
      });
      mod = require('../ConfigProvider');
      ConfigProvider = mod.default;
    });
    it('should change runtime locale, set theme, set primary theme color with default value', () => {
      const { getRunTimeLocale } = mod;
      const wrapper = shallow(
        <ConfigProvider>
          <></>
        </ConfigProvider>,
      );
      const actual = getRunTimeLocale();
      expect(actual).toEqual({});
      expect(mSetTheme).toBeCalledWith('light');
      expect(mSetPrimaryColor).toBeCalledWith('#00bc70');
      expect(wrapper.prop('value')).toEqual({
        locale: {},
        theme: 'light',
        primaryColor: '#00bc70',
      });
    });

    it('should accept only one child(a React element)', () => {
      expect(() =>
        shallow(
          <ConfigProvider>
            <div />
            <div />
          </ConfigProvider>,
        ),
      ).toThrowError('React.Children.only expected to receive a single React element child.');
    });

    it('should get context from ConfigContext provider', () => {
      const wrapper = mount(
        <ConfigProvider>
          <TestComponent />
        </ConfigProvider>,
      );
      expect(wrapper.find('div').text()).toEqual('light');
    });

    it('should get context from ConfigContext provider', () => {
      const wrapper = mount(
        <ConfigProvider>
          <ConfigContext.Consumer>{(value) => <div>{value.theme}</div>}</ConfigContext.Consumer>
        </ConfigProvider>,
      );
      expect(wrapper.find('div').text()).toEqual('light');
    });
  });
});
