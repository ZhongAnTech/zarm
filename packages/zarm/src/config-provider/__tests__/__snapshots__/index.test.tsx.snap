// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider snapshot renders correctly with en_US 1`] = `
<div
  class="za-alert za-alert--radius"
/>
`;

exports[`ConfigProvider snapshot renders correctly with zh_CN 1`] = `
<div
  class="za-alert za-alert--radius"
/>
`;

exports[`setTheme should set dark theme 1`] = `"--theme-primary-lighter: #303030; --color-text: rgba(255, 255, 255, 0.85); --color-text-inverse: rgba(255, 255, 255, 0.8); --color-text-placeholder: rgba(255, 255, 255, 0.3); --color-text-disabled: #666; --border-color: #3a3b3d; --opacity-mask: 0.7; --activity-indicator-path-color: var(--border-color); --action-sheet-background: #1b1c1e; --action-sheet-active-background: #363738; --alert-button-background: #2b2c2d; --alert-button-active-background: #363738; --button-default-background: #393939; --button-default-border: transparent; --button-default-color: rgba(255, 255, 255, 0.85); --button-default-active-background: rgba(255, 255, 255, 0.2); --button-default-active-border: transparent; --button-default-active-color: #fff; --calendar-background: transparent; --calendar-week-bar-background: #303030; --cell-background: #222; --cell-background-active: #363738; --cell-arrow-color: #666; --cell-arrow-disabled-color: #333; --checkbox-background: transparent; --checkbox-border-color: #58595d; --checkbox-disabled-background: #555; --checkbox-disabled-color: rgba(255, 255, 255, 0.3); --collapse-arrow-color: #666; --collapse-arrow-disabled-color: #333; --confirm-button-background: #2b2c2d; --confirm-button-active-background: #363738; --keyboard-background: #000; --keyboard-item-background: rgba(255, 255, 255, 0.1); --loading-background: #2b2c2d; --modal-background: #2b2c2d; --modal-close-color: rgba(255, 255, 255, 0.3); --modal-close-active-color: rgba(255, 255, 255, 0.65); --nav-bar-color: #1b1c1e; --panel-body-background: #1b1c1e; --picker-background: #000; --picker-header-background: #1b1c1e; --picker-mask-background-start: rgba(0, 0, 0, 0.4); --picker-mask-background-end: rgba(0, 0, 0, 0.8); --progress-background: var(--border-color); --radio-background: transparent; --radio-border-color: #58595d; --radio-disabled-background: #555; --radio-disabled-color: rgba(255, 255, 255, 0.3); --search-bar-background: #1b1c1e; --search-bar-inner-background: #000; --slider-line-dot-color: var(--border-color); --switch-background: #38393d; --stack-picker-background: #000; --stack-picker-shadow: none; --stepper-input-background: #000; --tabbar-background: #1b1c1e; --toast-background: #2b2c2d; --tooltip-background: #5b5c60;"`;

exports[`setTheme should set light theme 1`] = `""`;
